File: 130. Surrounded Regions/130. Surrounded Regions.py
class Solution:
    def solve(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        m,n=len(board),len(board[0])
        direc=[(0,1),(0,-1),(-1,0),(1,0)]
        safe=set()

        def dfs(r,c):
            if (r,c) in safe or board[r][c] != "O":
                return
            
            safe.add((r,c))

            for dr,dc in direc:
                nr=r+dr
                nc=c+dc

                if not(0<=nr<m and 0<=nc<n):
                    continue
                dfs(nr,nc)
        
        for r in range(m):
            dfs(r,0)
            dfs(r,n-1)
        
        for c in range(n):
            dfs(0,c)
            dfs(m-1,c)
        
        for r in range(m):
            for c in range(n):
                if (r,c) not in safe:
                    board[r][c]="X"

        


        

File: 130. Surrounded Regions/130. Surrounded Regions.cpp
class Solution {
public:

    struct pair_hash {
        size_t operator()(const pair<int,int>& p) const {
            return hash<int>()(p.first) ^ hash<int>()(p.second);
        }
    };

    unordered_set<pair<int,int>,pair_hash> safe = {};
    vector<pair<int,int>> direction = {{0,1},{0,-1},{-1,0},{1,0}};

    void dfs(vector<vector<char>>& board,int r, int c){
        if (board[r][c] != 'O' || safe.contains({r,c})) {
            return;
        }

        safe.insert({r,c});

        for(auto dir : direction){
            int nr = r + dir.first;
            int nc = c + dir.second;

            if (nr < 0 || nr >= board.size() ||
                nc < 0 || nc >= board[0].size()) {
                continue;
            }
            if(board[nr][nc]=='O'){
                dfs(board,nr,nc);
            }

        }

    }
    void solve(vector<vector<char>>& board) {
        

        //l and r side
        for(int r=0; r< board.size(); r++){
            dfs(board,r,0);
            dfs(board,r,board[0].size()-1);
        }
        
        //top bot side
        for(int c=0; c< board[0].size(); c++){
            dfs(board,0,c);
            dfs(board,board.size()-1,c);
        }

        for(int i=0; i<board.size();i++){
            for(int j=0; j<board[0].size();j++){
                if (safe.contains({i,j})){
                    continue;
                }
                else{
                    board[i][j]='X';
                }
            }
        }

        
    }
};